mixer-completion/cmd/completion.go

```diff
@@ -0,0 +1,68 @@
+// Copyright © 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"errors"
+	"os"
+
+	mixerCmd "github.com/clearlinux/mixer-tools/mixer/cmd"
+	"github.com/spf13/cobra"
+)
+
+// CompletionCmd represents the base command for mixer-completion
+var CompletionCmd = &cobra.Command{
+	Use:   "mixer-completion",
+	Short: "Generate autocomplete files for mixer",
+	Long: `Generates autocomplete files for mixer. If no arguments are passed, or if
+'bash' is passed, a bash completion file is written to /etc/bash_completion.d/.
+If 'zsh' is passed, a zsh completion file is written to
+/usr/share/zsh/site-functions/.
+
+These files allow you to type, for example:
+  mixer bun[TAB] -> mixer bundle a[TAB] -> mixer bundle add
+
+Note that this command must be run as root.`,
+	Args:      cobra.OnlyValidArgs,
+	ValidArgs: []string{"bash", "zsh"},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) > 1 {
+			return errors.New("mixer-completion takes at most one argument")
+		}
+
+		var shell string
+		if len(args) > 0 {
+			shell = args[0]
+		} else {
+			shell = "bash"
+		}
+
+		var err error
+		switch shell {
+		case "bash":
+			if err = os.MkdirAll("/etc/bash_completion.d", 0755); err != nil {
```

Comment:
I went with `0755` here (and below), which is what this directory is on my host system. Anyone have objection?

Reply:
If this is something that we install then clear seems to use

     /usr/share/bash-completion/completions

Reply:
I considered putting it there, but I read a number of places (like [this one](https://lists.fedoraproject.org/pipermail/devel/2011-June/153012.html) about Fedora from 2011) that claim `/usr/share/bash-completion/` is where the bash-completion package itself installs some completions (and symlinks to `/etc/bash_completion.d/` to enable/disable), and it is this second location where completions are actually loaded from, and where custom completions are to be installed from packages.

It's entirely possible that's no longer true, or that it was never true for Clear.  The vast majority of places just tell you to put things in `/etc/bash_completion.d/` and don't mention anywhere else.  I did not try putting things in `/usr/share/bash-completion/`, so I don't know if it would pick them up from there, but it does successfully pick them up from `/etc/bash_completion.d/`

Reply:
The point is that for packaging purposes and because we are stateless, we *never* ship anything in /etc. see packages/bash/stateless.patch in the clear source tree for example where we pick up /usr/share/defaults/etc/bash.bashrc rather than /etc/bash.bashrc.